haystack/core/pipeline/base.py

```diff
@@ -1127,6 +1118,60 @@ def _component_has_enough_inputs_to_run(self, name: str, inputs: Dict[str, Dict[
         current_inputs = inputs[name].keys()
         return expected_inputs == current_inputs
 
+    def _break_supported_cycles_in_graph(self) -> Tuple[networkx.MultiDiGraph, Dict[str, List[List[str]]]]:
+        """
+        Utility function to remove supported cycles in the Pipeline's graph.
+
+        The way the Pipeline execution works we can only break connections in cycles that
+        have a Variadic or GreedyVariadic type or a default value.
+
+        This will raise a PipelineRuntimeError if we there are cycles that can't be broken.
+        That is bound to happen when all the inputs in a cycle are mandatory.
+
+        If the Pipeline's graph doesn't have any cycle it will just return that graph and an empty dictionary.
+
+        :return:
+            A tuple containing:
+                * A copy of the Pipeline's graph without cycles
+                * A dictionary of Component's names and all the cycles they were part of
+        """
+        if networkx.is_directed_acyclic_graph(self.graph):
+            return self.graph, {}
+
+        temp_graph: networkx.MultiDiGraph = self.graph.copy()
+        cycles: List[List[str]] = networkx.recursive_simple_cycles(self.graph)
+        edges_removed: Dict[str, List[str]] = {}
+        # This keeps track of all the cycles that a component is part of.
+        components_in_cycles: Dict[str, List[List[str]]] = {}
+        for cycle in cycles:
+            for comp in cycle:
+                if comp not in components_in_cycles:
+                    components_in_cycles[comp] = []
+                components_in_cycles[comp].append(cycle)
+
+            for sender_comp, receiver_comp in zip(cycle, cycle[1:] + cycle[:1]):
```

Comment:
A comment to explain the second argument to `zip`. And while we're at it, a high-level description of what this algorithm does would be helpful to follow the code (it can be either part of the docstring or as an function-internal comment).